src/librustc/traits/project.rs

```diff
@@ -1320,23 +1320,10 @@ fn assoc_ty_def<'cx, 'gcx, 'tcx>(
     let trait_def_id = selcx.tcx().impl_trait_ref(impl_def_id).unwrap().def_id;
     let trait_def = selcx.tcx().trait_def(trait_def_id);
 
-    if !trait_def.is_complete(selcx.tcx()) {
```

Comment:
Thanks, @eddyb! 
How does the following look to you?
```rust
fn assoc_ty_def<'cx, 'gcx, 'tcx>(
    selcx: &SelectionContext<'cx, 'gcx, 'tcx>,
    impl_def_id: DefId,
    assoc_ty_name: ast::Name)
    -> Option<specialization_graph::NodeItem<ty::AssociatedItem>>
{
    let tcx = selcx.tcx();
    let trait_def_id = tcx.impl_trait_ref(impl_def_id).unwrap().def_id;
    let trait_def = tcx.trait_def(trait_def_id);

    // Check for a cycle
    let specialization_graph_complete =
        ty::queries::specialization_graph_of::try_get(tcx, DUMMY_SP, self.def_id).is_ok();

    if !specialization_graph_complete {
        let impl_node = specialization_graph::Node::Impl(impl_def_id);
        for item in impl_node.items(selcx.tcx()) {
            if item.kind == ty::AssociatedKind::Type && item.name == assoc_ty_name {
                return Some(specialization_graph::NodeItem {
                    node: specialization_graph::Node::Impl(impl_def_id),
                    item: item,
                });
            }
        }
        None
    } else {
        trait_def
            .ancestors(tcx, impl_def_id)
            .defs(tcx, assoc_ty_name, ty::AssociatedKind::Type)
            .next()
    }
}
```
(I would clean this up of course and add some comments)